README.md
## Django Advanced Filters

| Branch | Build | Coverage | PyPI | Gitter |
| ------ | ----- | -------- | ---- | ------ |
| Master | [![Build Status][5]][7] | [![Coverage Status][8]][10] | [![PyPI][1]][2] | [![Gitter][3]][4] |
| Develop | [![Build Status][6]][7] | [![Coverage Status][9]][11] | 🔺 | 🔺 |


[1]: https://img.shields.io/pypi/pyversions/django-advanced-filters.svg
[2]: https://pypi.python.org/pypi/django-advanced-filters
[3]: https://badges.gitter.im/Join%20Chat.svg
[4]: https://gitter.im/modlinltd/django-advanced-filters?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge
[5]: https://travis-ci.org/modlinltd/django-advanced-filters.svg?branch=master
[6]: https://travis-ci.org/modlinltd/django-advanced-filters.svg?branch=develop
[7]: https://travis-ci.org/modlinltd/django-advanced-filters
[8]: https://coveralls.io/repos/modlinltd/django-advanced-filters/badge.svg?branch=master
[9]: https://coveralls.io/repos/modlinltd/django-advanced-filters/badge.svg?branch=develop
[10]: https://coveralls.io/github/modlinltd/django-advanced-filters?branch=master
[11]: https://coveralls.io/github/modlinltd/django-advanced-filters?branch=develop


A django ModelAdmin mixin which adds advanced filtering abilities to the admin.

Mimics the advanced search feature in [VTiger](https://www.vtiger.com/),
[see here for more info](https://wiki.vtiger.com/index.php/Create_Custom_Filters)

![screenshot](https://raw.githubusercontent.com/modlinltd/django-advanced-filters/develop/screenshot.png "Creating via a modal")

For release notes, see [Changelog](https://raw.githubusercontent.com/modlinltd/django-advanced-filters/develop/CHANGELOG.md)

## Requirements

* Django >= 1.5  (Django 1.5 - 1.8 on Python 2/3/PyPy2)
* django-easy-select2 == 1.2.5
* django-braces == 1.4.0
* simplejson == 3.6.5


## Installation & Set up

1. Install from pypi: `pip install django-advanced-filters`
2. Add both `'advanced_filters'` and `'easy_select2'` to `INSTALLED_APPS`.
3. Add `url(r'^advanced_filters/', include('advanced_filters.urls'))` to your project's urlconf.
4. Run `python manage.py syncdb`

## Integration Example

Extending a ModelAdmin is pretty straightforward:

    class ProfileAdmin(AdminAdvancedFiltersMixin, models.ModelAdmin):
        list_filter = ('name', 'language', 'ts')   # simple list filters

        # select from these fields in the advanced filter creation form
        advanced_filter_fields = (
            'name', 'language', 'ts'
            # even use related fields as lookup fields
            'country__name', 'posts__title', 'comments__content'
        )

Adding a new advanced filter (see below) will display a new list filter
named "Advanced filters" which will list all the filter the currently
logged in user is allowed to use (by default only those he/she created).

### Custom naming of fields

Initially, each field in `advanced_filter_fields` is resolved into an actual
model field. That field's verbose_name attribute is then used as the text
of the displayed option. While uncommon, it occasionally makes sense to
use a custom name, especially when following a relationship, as the context
then changes.

For example, when a profile admin allows filtering by a user name as well as
a sales representative name, it'll get confusing:

    class ProfileAdmin(AdminAdvancedFiltersMixin, models.ModelAdmin):
      advanced_filter_fields = ('name', 'sales_rep__name')

In this case the field options will both be named "name" (by default).

To fix this, use custom naming:


    class ProfileAdmin(AdminAdvancedFiltersMixin, models.ModelAdmin):
      advanced_filter_fields = ('name', ('sales_rep__name', 'assigned rep'))

Now, you will get two options, "name" and "assigned rep".


## Adding new advanced filters

By default the mixin uses a template which extends django's built-in
`change_list` template. This template is based off of grapelli's
fork of this template (hence the 'grp' classes and funny looking javascript).

The default template also uses the superb [magnificPopup](dimsemenov/Magnific-Popup)
which is currently bundled with the application.

Regardless of the above, you can easily write your own template which uses
context variables `{{ advanced_filters }}` and
`{{ advanced_filters.formset }}`, to render the advanced filter creation form.

## Structure

Each advanced filter has only a couple of required fields when constructed
with the form; namely the title and a formset (consisting of a form for each
sub-query or rule of the filter query).

Each form in the formset requires the following fields:
`field`, `operator`, `value`

And allows the optional `negate` and `remove` fields.

Let us go over each of the fields in a rule fieldset.

### Field

The list of all available fields for this specific instance of the ModelAdmin
as specific by the [`advanced_filter_fields` property.](#integration-example)

#### The OR field

`OR` is an additional field that is added to every rule's available fields.

It allows constructing queries with [OR statements](https://docs.djangoproject.com/en/dev/topics/db/queries/#complex-lookups-with-q-objects). You can use it by creating an "empty" rule with this field "between" a set of 1 or more rules.

### Operator

Query field suffixes which specify how the `WHERE` query will be constructed.

The currently supported are as follows:
`iexact`, `icontains`, `iregex`, `range`, `isnull`, `istrue` and `isfalse`

For more detail on what they mean and how they function, see django's
[documentation on field lookups](https://docs.djangoproject.com/en/dev/ref/models/querysets/#field-lookups).

### Value

The value which the specific sub-query will be looking for, i.e the
value of the field specified above, or in django query syntax: `.filter(field=value)`

### Negate

A boolean (check-box) field to specify whether this rule is to be negated,
effectively making it a "exclude" sub-query.

### Remove

Similarly to other [django formsets](https://docs.djangoproject.com/en/dev/topics/forms/formsets/),
used to remove the formset on submit.


## Editing previously created advanced filters

The `AdvancedFilterAdmin` class (a subclass of `ModelAdmin`) is provided
and registered with `AdvancedFilter` in admin.py module.

The model's change_form template is overridden from grapelli's/django's
standard template, to mirror the add form modal as closely as possible.

*Note:* currently, adding new filters from the ModelAdmin change page
is not supported.

## Query Serialization

**TODO:** write a few words on how serialization of queries is done.


## Model correlation

Since version 1.0, The underlying `AdvancedFilter` model instances are tightly
coupled with a specific model (using the app_label.Name model name),
for which admin changelist they are to used and created in.

This change has a few benefits:

1. Admin mixin can be used with multiple `ModelAdmin` classes while performing
   specific query serialization and field validation that are at the base of the
   filter functionality.

2. Allows users to edit previously created filters outside of
   the context of a changelist, as we do in the [`AdvancedFilterAdmin`](#editing-previously-created-advanced-filters).

3. Limit the `AdvancedListFilters` to limit queryset (and thus, the underlying
   options) to a specified model.

Note: Since we are at the early stages of development I have skipped the
South / 1.7 schema (new field) and data migrations (add specific model to all
existing instances of AdvancedFilter model) migrations. Though this shouldn't
be too difficult to do, if the need arises I can add migration examples.

## Views

The GetFieldChoices view is required to dynamically (using javascript) fetch a
list of valid field choices when creating/changing an `AdvancedFilter`.


## TODO

* ~~Add more tests (specifically the admin and view parts)~~
* ~~Add packaging (setup.py, etc...)~~
* ~~Add edit/view functionality to the filters~~
* Add permission user/group selection functionality to the filter form
* Allow toggling of predefined templates (grappelli / vanilla django admin), and front-end features.
* Support more (newer) python/django versions


CHANGELOG.md
# Changelog

## 1.0.2 - A Better Future

This release features better test coverage and support for Django 1.9.

### Bugs
   - stretch formset table to the modal container width
   - toggle advanced `vendor/jquery` dir according to Django version
   - retain support older Django versions
   - clean up legacy tags in templates

### Tests
   - add admin views tests
   - add Django 1.9 to test matrix
   - other minor improvements

### Docs
   - Improve README with a newer screenshot and pretty tables for badges

### Contributors:
   - Pavel Savchenko
   - Leonardo J. Caballero G
   - Schuyler Duveen

## 1.0.1 - A Public Release

### Bugs
   - proper support for py26 and py3X and different Django releases
   - avoid querying all instances for choices
   - resolve settings inside view and refine error handling

### Tests
   - add doctests to the `form_helpers`
   - add tests for `forms`
   - add test case `views.TestGetFieldChoicesView`
   - setup.py/travis: add `test-reqs.txt` as extras_require
   - refactor testing to use `py.test` and run `tox` from `setup.py`
   - travis: use latest version of each Django release

### Docs:
   - `README`: explain what we test against

## 1.0 - First contact

#### Major changes
* Add a new (required) field [`AdvancedFilter.model`](README.md#model-correlation)
* Add parsing query dict into initialized formsets (allows for [editing existing instance](README.md#editing-previously-created-advanced-filters)).
* Add [`AdvancedFilterAdmin`](#editing-previously-created-advanced-filters) for actually accessing and [editing existing `AdvancedFilter` instances](README.md#editing-previously-created-advanced-filters).
* Use [Select2](https://github.com/asyncee/django-easy-select2) and an AJAX view to
dynamically populate [`field` options](README.md#fields).
* Add proper support for nested serialization of queries.

#### Minor changes
* Implement more [`operators`](README.md#operators) (`isnull`, `istrue` and `isfalse`)
* Allow [custom verbose naming of fields in advanced_filter_fields](README.md#custom-naming-of-fields)
* Add helper methods to the model to hide (and decouple) core serialization functionality from users.
* Strip whitespace in field values validation
* Setup and packaging (`setup.py`/`MANIFEST.in`)
* Hide `QSerializer` calling logic in the model
* Allow modifying `advanced_filter_form` property (defaults to `AdvancedFilterForm`)
* Correct documentation regarding position of mixin in subclass (issue #1)


advanced_filters/forms.py
from datetime import datetime as dt
from pprint import pformat
import logging
import operator

from django import forms

try:
    from django.apps import apps
    get_model = apps.get_model
except ImportError:
    # django < 1.7 support
    from django.db.models import get_model

from django.conf import settings
from django.contrib import admin

try:
    from django.contrib.admin.utils import get_fields_from_path
except ImportError:
    # django < 1.7 support
    from django.contrib.admin.util import get_fields_from_path

from django.db.models import Q, FieldDoesNotExist
from django.db.models.fields import DateField
from django.forms.formsets import formset_factory, BaseFormSet
from django.templatetags.static import static
from django.utils.functional import cached_property
from django.utils.translation import ugettext_lazy as _
from django.utils.six.moves import range, reduce
from django.utils.text import capfirst

import django

from easy_select2.widgets import SELECT2_WIDGET_JS, SELECT2_CSS

from .models import AdvancedFilter
from .form_helpers import CleanWhiteSpacesMixin,  VaryingTypeCharField


# django < 1.9 support
USE_VENDOR_DIR = django.VERSION >= (1, 9)
logger = logging.getLogger('advanced_filters.forms')


class AdvancedFilterQueryForm(CleanWhiteSpacesMixin, forms.Form):
    """ Build the query from field, operator and value """
    OPERATORS = (
        ("iexact", _("Equals")),
        ("icontains", _("Contains")),
        ("iregex", _("One of")),
        ("range", _("DateTime Range")),
        ("isnull", _("Is NULL")),
        ("istrue", _("Is TRUE")),
        ("isfalse", _("Is FALSE")),
    )

    FIELD_CHOICES = (
        ("_OR", _("Or (mark an or between blocks)")),
    )

    field = forms.ChoiceField(required=True, widget=forms.Select(
        attrs={'class': 'query-field'}))
    operator = forms.ChoiceField(
        required=True, choices=OPERATORS, initial="iexact",
        widget=forms.Select(attrs={'class': 'query-operator'}))
    value = VaryingTypeCharField(required=True, widget=forms.TextInput(
        attrs={'class': 'query-value'}))
    value_from = forms.DateTimeField(widget=forms.HiddenInput(
        attrs={'class': 'query-dt-from'}), required=False)
    value_to = forms.DateTimeField(widget=forms.HiddenInput(
        attrs={'class': 'query-dt-to'}), required=False)
    negate = forms.BooleanField(initial=False, required=False)

    def _build_field_choices(self, fields):
        """
        Iterate over passed model fields tuple and update initial choices.
        """
        return tuple(sorted(
            [(fquery, capfirst(fname)) for fquery, fname in fields.items()],
            key=lambda f: f[1].lower())
        ) + self.FIELD_CHOICES

    def _build_query_dict(self, formdata=None):
        """
        Take submitted data from form and create a query dict to be
        used in a Q object (or filter)
        """
        if self.is_valid() and formdata is None:
            formdata = self.cleaned_data
        key = "{field}__{operator}".format(**formdata)
        if formdata['operator'] == "isnull":
            return {key: None}
        elif formdata['operator'] == "istrue":
            return {formdata['field']: True}
        elif formdata['operator'] == "isfalse":
            return {formdata['field']: False}
        return {key: formdata['value']}

    @staticmethod
    def _parse_query_dict(query_data, model):
        """
        Take a list of query field dict and return data for form initialization
        """
        if query_data['field'] == '_OR':
            query_data['operator'] = 'iexact'
            return query_data

        parts = query_data['field'].split('__')
        if len(parts) < 2:
            field = parts[0]
        else:
            if parts[-1] in dict(AdvancedFilterQueryForm.OPERATORS).keys():
                field = '__'.join(parts[:-1])
            else:
                field = query_data['field']

        query_data['field'] = field
        mfield = get_fields_from_path(model, query_data['field'])
        if not mfield:
            raise Exception('Field path "%s" could not be followed to a field'
                            ' in model %s', query_data['field'], model)
        else:
            mfield = mfield[-1]  # get the field object

        if query_data['value'] is None:
            query_data['operator'] = "isnull"
        elif query_data['value'] is True:
            query_data['operator'] = "istrue"
        elif query_data['value'] is False:
            query_data['operator'] = "isfalse"
        else:
            if isinstance(mfield, DateField):
                # this is a date/datetime field
                query_data['operator'] = "range"  # default
            else:
                query_data['operator'] = "iexact"  # default

        if isinstance(query_data.get('value'),
                      list) and query_data['operator'] == 'range':
            dtfrom = dt.fromtimestamp(query_data.get('value_from', 0))
            dtto = dt.fromtimestamp(query_data.get('value_to', 0))
            query_data['value'] = ','.join([dtfrom.strftime('%Y-%m-%d'),
                                            dtto.strftime('%Y-%m-%d')])

        return query_data

    def set_range_value(self, data):
        """
        Validates date range by parsing into 2 datetime objects and
        validating them both.
        """
        dtfrom = data.pop('value_from')
        dtto = data.pop('value_to')
        if dtfrom is dtto is None:
            self.errors['value'] = ['Date range requires values']
            raise forms.ValidationError([])
        data['value'] = (dtfrom, dtto)

    def clean(self):
        cleaned_data = super(AdvancedFilterQueryForm, self).clean()
        if cleaned_data.get('operator') == "range":
            if ('value_from' in cleaned_data and
                    'value_to' in cleaned_data):
                self.set_range_value(cleaned_data)
        return cleaned_data

    def make_query(self, *args, **kwargs):
        """ Returns a Q object from the submitted form """
        query = Q()  # initial is an empty query
        query_dict = self._build_query_dict(self.cleaned_data)
        if 'negate' in self.cleaned_data and self.cleaned_data['negate']:
            query = query & ~Q(**query_dict)
        else:
            query = query & Q(**query_dict)
        return query

    def __init__(self, model_fields={}, *args, **kwargs):
        super(AdvancedFilterQueryForm, self).__init__(*args, **kwargs)
        self.FIELD_CHOICES = self._build_field_choices(model_fields)
        self.fields['field'].choices = self.FIELD_CHOICES
        if not self.fields['field'].initial:
            self.fields['field'].initial = self.FIELD_CHOICES[0]


class AdvancedFilterFormSet(BaseFormSet):
    """ """
    fields = ()
    extra_kwargs = {}

    def __init__(self, *args, **kwargs):
        self.model_fields = kwargs.pop('model_fields', {})
        super(AdvancedFilterFormSet, self).__init__(*args, **kwargs)
        if self.forms:
            form = self.forms[0]
            self.fields = form.visible_fields()

    @property
    def empty_form(self):
        form = self.form(
            model_fields=self.model_fields,
            auto_id=self.auto_id,
            prefix=self.add_prefix('__prefix__'),
            empty_permitted=True,
        )
        self.add_fields(form, None)
        return form

    def _construct_forms(self):
        # not strictly required, but Django 1.5 calls this on init
        # django == 1.5 support
        self.forms = []
        for i in range(min(self.total_form_count(), self.absolute_max)):
            self.forms.append(self._construct_form(
                i, model_fields=self.model_fields))

    @cached_property
    def forms(self):
        # override the original property to include `model_fields` argument
        forms = [self._construct_form(i, model_fields=self.model_fields)
                 for i in range(self.total_form_count())]
        forms.append(self.empty_form)  # add initial empty form
        return forms

AFQFormSet = formset_factory(
    AdvancedFilterQueryForm, formset=AdvancedFilterFormSet,
    extra=1, can_delete=True)

AFQFormSetNoExtra = formset_factory(
    AdvancedFilterQueryForm, formset=AdvancedFilterFormSet,
    extra=0, can_delete=True)


class AdvancedFilterForm(CleanWhiteSpacesMixin, forms.ModelForm):
    """ Form to save/edit advanced filter forms """
    class Meta:
        model = AdvancedFilter
        fields = ('title',)

    class Media:
        required_js = [static('admin/js/%sjquery.min.js' %
                       ('vendor/jquery/' if USE_VENDOR_DIR else '')),
                       static('orig_inlines%s.js' %
                       ('' if settings.DEBUG else '.min')),
                       static('magnific-popup/jquery.magnific-popup.js'),
                       static('advanced-filters/advanced-filters.js'), ]
        js = SELECT2_WIDGET_JS + required_js
        css = {'screen': [static(SELECT2_CSS),
                          static('advanced-filters/advanced-filters.css'),
                          static('magnific-popup/magnific-popup.css')]}

    def get_fields_from_model(self, model, fields):
        """
        Iterate over given <field> names (in "orm query" notation) and find
        the actual field given the initial <model>.

        If <field> is a tuple of the format ('field_name', 'Verbose name'),
        overwrite the field's verbose name with the given name for display
        purposes.
        """
        model_fields = {}
        for field in fields:
                if isinstance(field, tuple) and len(field) == 2:
                    field, verbose_name = field[0], field[1]
                else:
                    try:
                        model_field = get_fields_from_path(model, field)[-1]
                        verbose_name = model_field.verbose_name
                    except (FieldDoesNotExist, IndexError, TypeError) as e:
                        logger.warn("AdvancedFilterForm: skip invalid field "
                                    "- %s", e)
                        continue
                model_fields[field] = verbose_name
        return model_fields

    def __init__(self, *args, **kwargs):
        model_admin = kwargs.pop('model_admin', None)
        instance = kwargs.get('instance')
        extra_form = kwargs.pop('extra_form', False)
        # TODO: allow all fields to be determined by model
        filter_fields = kwargs.pop('filter_fields', None)
        if model_admin:
            self._model = model_admin.model
        elif instance and instance.model:
            # get existing instance model
            self._model = get_model(*instance.model.split('.'))
            try:
                model_admin = admin.site._registry[self._model]
            except KeyError:
                logger.debug('No ModelAdmin registered for %s', self._model)
        else:
            raise Exception('Adding new AdvancedFilter from admin is '
                            'not supported')

        self._filter_fields = filter_fields or getattr(
            model_admin, 'advanced_filter_fields', ())

        super(AdvancedFilterForm, self).__init__(*args, **kwargs)

        # populate existing or empty forms formset
        data = None
        if len(args):
            data = args[0]
        elif kwargs.get('data'):
            data = kwargs.get('data')
        self.initialize_form(instance, self._model, data, extra_form)

    def clean(self):
        cleaned_data = super(AdvancedFilterForm, self).clean()
        if not self.fields_formset.is_valid():
            logger.debug(
                "Errors validating advanced query filters: %s",
                pformat([(f.errors, f.non_field_errors())
                         for f in self.fields_formset.forms]))
            raise forms.ValidationError("Error validating filter forms")
        cleaned_data['model'] = "%s.%s" % (self._model._meta.app_label,
                                           self._model._meta.object_name)
        return cleaned_data

    @property
    def _non_deleted_forms(self):
        forms = []
        for form in self.fields_formset.forms:
            if form in self.fields_formset.deleted_forms:
                continue  # skip deleted forms when generating query
            forms.append(form)
        return forms

    def generate_query(self):
        """ Reduces multiple queries into a single usable query """
        query = Q()
        ORed = []
        for form in self._non_deleted_forms:
            if not hasattr(form, 'cleaned_data'):
                continue
            if form.cleaned_data['field'] == "_OR":
                ORed.append(query)
                query = Q()
            else:
                query = query & form.make_query()
        if ORed:
            if query:  # add last query for OR if any
                ORed.append(query)
            query = reduce(operator.or_, ORed)
        return query

    def initialize_form(self, instance, model, data=None, extra=None):
        """ Takes a "finalized" query and generate it's form data """
        model_fields = self.get_fields_from_model(model, self._filter_fields)

        forms = []
        if instance:
            for field_data in instance.list_fields():
                forms.append(
                    AdvancedFilterQueryForm._parse_query_dict(
                        field_data, model))

        formset = AFQFormSetNoExtra if extra is None else AFQFormSet
        self.fields_formset = formset(
            data=data,
            initial=forms or None,
            model_fields=model_fields
        )

    def save(self, commit=True):
        self.instance.query = self.generate_query()
        self.instance.model = self.cleaned_data.get('model')
        return super(AdvancedFilterForm, self).save(commit)
